File: WeeklyRotators/UpdateActiveWeeklyRotators.cs
using Microsoft.Data.SqlClient;
using System.Text;
using Serilog;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

public class Activity{
    // Property for the type of activity
    public string? ActivityType { get; set; }

    // Property for the hash associated with the activity
    public string? Hash { get; set; }

    // Property for the sequence number
    public int? Sequence { get; set; }

    public string? Name { get; set; }
    
    // Constructor to initialize the properties
    public Activity(string activityType, string hash, int sequence, string name){
        ActivityType = activityType;
        Hash = hash;
        Sequence = sequence;
        Name = name;
    }
    
}

public class RotatorEntry{
    public string? RotatorType { get; set; }
    public int? MaxSequence { get; set; }
}

public class UpdateActiveWeeklyRotators{
    public static void UpdateTable(string Server, string Database, string UserId, string Password){
        // Create MS SQL Server connection
        string sqlConnectionString = $"Server={Server};Database={Database};TrustServerCertificate=True;Uid={UserId};Pwd={Password};";
        Log.Information(sqlConnectionString);
        string tableName = "ActiveWeeklyRotatorsTable";
 
        using (SqlConnection msSqlConnection = new SqlConnection(sqlConnectionString)){
            try{
                msSqlConnection.Open();
                bool doesTableExist = TableExists(connection:msSqlConnection, tableName:tableName);
                if(doesTableExist == false){
                    Log.Information($"{tableName} not found");
                    using (SqlCommand createTableCommand = new SqlCommand($"CREATE TABLE {tableName} (NAME VARCHAR(MAX) , RotatorList VARCHAR(MAX))", msSqlConnection)) {
                        createTableCommand.ExecuteNonQuery();
                    }
                }
                // If the ActiveWeeklyRotators does not exist it will set the table to the start of the rotator sequence
                bool doesNameExistsInColumn = NameExistsInColumn(connection:msSqlConnection, tableName:tableName, columnName:"NAME", nameToCheck:"ActiveWeeklyRotators");
                if(doesNameExistsInColumn == false){
                    string valueForNameColumn = "ActiveWeeklyRotators";
                    Log.Information($"{valueForNameColumn} name in column was not found. Will run ResetAndPopulateWeeklyRotatorsTable method");
                    ResetAndPopulateWeeklyRotatorsTable(msSqlConnection:msSqlConnection, tableName:tableName, valueForNameColumn:valueForNameColumn);
                }else if(doesNameExistsInColumn == true){
                    Log.Information("The ActiveWeeklyRotatorsTableExist and it is not empty....Attempting to set new rotators");
                    string valueForNameColumn = "ActiveWeeklyRotators";
                    string valueForRotatorListColumn = "RotatorList";
                    string? currentActiveWeeklyRotatorsString = "";

                    // This Query gets the Active Weekly Rotators String from SQL Server and sets it to a variable
                    using (SqlCommand findActiveWeeklyRotatorsQuery = new SqlCommand($"SELECT {valueForRotatorListColumn} FROM {tableName} WHERE NAME = @NameValue", msSqlConnection)){
                        findActiveWeeklyRotatorsQuery.Parameters.AddWithValue("@NameValue", valueForNameColumn);
                        using (SqlDataReader reader = findActiveWeeklyRotatorsQuery.ExecuteReader()){
                            if (reader.Read()){
                                currentActiveWeeklyRotatorsString = reader[valueForRotatorListColumn].ToString();
                            }
                        }
                    }
                    Log.Information("Got the Active Weekly Rotators String");

                    // This will set a default weekly rotators table and end UpdateActiveWeeklyRotatorsMethod if there is something wrong with the table
                    if(currentActiveWeeklyRotatorsString == null || currentActiveWeeklyRotatorsString == ""){
                        Log.Information("The Active Weekly Rotators String is either empty or null");
                        ResetAndPopulateWeeklyRotatorsTable(msSqlConnection:msSqlConnection, tableName:tableName, valueForNameColumn:valueForNameColumn);
                        return; 
                    }

                    // Remove Square brackets from string
                    string trimmedCurrentActiveWeeklyRotatorsString = currentActiveWeeklyRotatorsString.Trim('[', ']');
                    // Split the string by commas
                    string[] currentActiveWeeklyRotatorsStringArray = trimmedCurrentActiveWeeklyRotatorsString.Split(',');
                    // Convert the string array to a list of integers
                    List<long> currentActiveWeeklyRotatorsList = currentActiveWeeklyRotatorsStringArray
                        .Select(s => long.Parse(s.Trim())) // Parse string to long
                        .ToList();
                    
                    int? currentActivitySequence = null;
                    string? currentActivityRotatorType = "";
                    List<Activity> newActiveWeeklyRotatorsList = new List<Activity>();
                    List<RotatorEntry> rotatorEntries = GetMaxSequenceByRotatorType(msSqlConnection:msSqlConnection, tableName:"RotatorSchedule", rotatorTypeColumnName:"RotatorType");
                    // Loops through all the hashes in the current rotator list and returns the new active weekly rotator list
                    Log.Information("Looping through the currentActiveWeeklyRotatorsList");
                    foreach(long hash in currentActiveWeeklyRotatorsList){
                        // Query to get current Activity Sequence and RotatorType
                        using (SqlCommand findCurrentRotatorQuery = new SqlCommand($"SELECT Sequence,RotatorType,Hash FROM RotatorSchedule WHERE Hash = @Hash",msSqlConnection)) {
                            findCurrentRotatorQuery.Parameters.AddWithValue("@Hash", hash);
                            using (SqlDataReader reader = findCurrentRotatorQuery.ExecuteReader()){
                                if (reader.Read()){
                                    currentActivitySequence = Convert.ToInt32(reader["Sequence"]);
                                    currentActivityRotatorType = reader["RotatorType"].ToString();
                                }
                            }
                        }

                        // checks if rotatortype found is at the end of the sequence. If it is..it resets sequence to 1 else it adds 1 to sequence
                        int? newActivitySequence = null;
                        foreach (var entry in rotatorEntries){
                            if(entry.RotatorType == currentActivityRotatorType){
                                if(currentActivitySequence >= entry.MaxSequence){
                                    newActivitySequence = 1;
                                }else{
                                    newActivitySequence = currentActivitySequence + 1;
                                }
                            }
                        }

                        // Builds the Object Activity that is then placed into the AcitveWeeklyRotatorsList
                        using(SqlCommand findNewRotatorQuery = new SqlCommand($"SELECT Hash, RotatorType, Sequence, Name FROM RotatorSchedule WHERE Sequence = @NewActivitySequence AND RotatorType = @NewActivityRotatorType",msSqlConnection)){
                            findNewRotatorQuery.Parameters.AddWithValue("@NewActivitySequence", newActivitySequence);
                            findNewRotatorQuery.Parameters.AddWithValue("@NewActivityRotatorType", currentActivityRotatorType);
                            using (SqlDataReader reader = findNewRotatorQuery.ExecuteReader()) {
                                while (reader.Read()){
                                    Activity newActivity = new Activity(
                                        activityType: reader["RotatorType"].ToString(),
                                        hash: reader["Hash"].ToString(),
                                        sequence: Convert.ToInt32(reader["Sequence"]),
                                        name: reader["Name"].ToString()
                                    );
                                    newActiveWeeklyRotatorsList.Add(newActivity);
                                }
                            }
                        }
                    }

                    bool raid = false;
                    bool dungeon = false;
                    bool exoticMission = false;

                    foreach (Activity activity in newActiveWeeklyRotatorsList){
                        if (activity.ActivityType == "Raid"){
                            raid = true;
                        } else if(activity.ActivityType == "Dungeon"){
                            dungeon = true;
                        } else if(activity.ActivityType == "Exotic Mission"){
                            exoticMission = true;
                        }
                    }

                    string newActiveWeeklyRotatorsString = "";
                    if (raid == true && dungeon == true && exoticMission == true){
                        Log.Information("A new raid, dungeon, and exotic mission was found. Updating Active Weekly Rotators Table");
                        // Collect hashes into a list
                        var hashes = newActiveWeeklyRotatorsList
                            .Select(activity => activity.Hash) // Select hashes
                            .ToList(); // Convert to list

                        // Join hashes with commas and format with square brackets
                        newActiveWeeklyRotatorsString = "[" + string.Join(",", hashes) + "]";

                        using (SqlCommand updateRowCommand = new SqlCommand($"UPDATE {tableName} SET RotatorList = @RotatorList", msSqlConnection)){
                            updateRowCommand.Parameters.AddWithValue("@RotatorList", string.Join(",", newActiveWeeklyRotatorsString));
                            updateRowCommand.ExecuteNonQuery();
                        }
                    }else{
                        Log.Error("The code could not find one of the previous Raids, Dungeon or ExoticMission");
                        ResetAndPopulateWeeklyRotatorsTable(msSqlConnection:msSqlConnection, tableName:tableName, valueForNameColumn:valueForNameColumn);
                    }
                }
            }catch(Exception ex){
                Log.Error(ex.Message);
            }
        }

  
    }

    private static bool TableExists(SqlConnection connection, string tableName) {
        using (SqlCommand command = new SqlCommand($"SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = '{tableName}'", connection)) {
            return command.ExecuteScalar() != null;
        }
    }
    private static bool NameExistsInColumn(SqlConnection connection, string tableName, string columnName, string nameToCheck){
        string sqlQuery = $"SELECT 1 FROM {tableName} WHERE {columnName} = @NameToCheck";
        using (SqlCommand command = new SqlCommand(sqlQuery, connection)){
            command.Parameters.AddWithValue("@NameToCheck", nameToCheck);
            return command.ExecuteScalar() != null;
        }
    }
    private static void ResetAndPopulateWeeklyRotatorsTable(SqlConnection msSqlConnection, string tableName, string valueForNameColumn){
        Log.Information("Dropping Active Weekly Rotators Table");
        using (SqlCommand deleteTableCommand = new SqlCommand($"DROP TABLE {tableName}", msSqlConnection)){
            deleteTableCommand.ExecuteNonQuery();
        }

        Log.Information("Creating Active Weekly Rotators Table");
        using (SqlCommand createTableCommand = new SqlCommand($"CREATE TABLE {tableName} (NAME VARCHAR(MAX), RotatorList VARCHAR(MAX))", msSqlConnection)){
            createTableCommand.ExecuteNonQuery();
        }

        List<long> defaultWeeklyRotatorList = new List<long>{
            2122313384,
            2823159265,
            2668737148
        };

        string defaultWeeklyRotatorListString = "[" + string.Join(",", defaultWeeklyRotatorList) + "]";
        Log.Information("The Active Weekly Rotators Table was found but it's either empty or missing rotators or one the next sequence number was not found. Setting default.");

        using (SqlCommand addDefaultRowCommand = new SqlCommand($"INSERT INTO {tableName} (NAME, RotatorList) VALUES (@NAME, @RotatorList)", msSqlConnection)){
            addDefaultRowCommand.Parameters.AddWithValue("@NAME", valueForNameColumn);
            addDefaultRowCommand.Parameters.AddWithValue("@RotatorList", defaultWeeklyRotatorListString);
            addDefaultRowCommand.ExecuteNonQuery();
        }
    }

    private static List<RotatorEntry> GetMaxSequenceByRotatorType(SqlConnection msSqlConnection, string tableName, string rotatorTypeColumnName){
        var result = new List<RotatorEntry>();
        using (SqlCommand getMaxSequenceByRotatorTypeQuery = new SqlCommand($"SELECT {rotatorTypeColumnName}, MAX(Sequence) AS MaxSequence FROM {tableName} GROUP BY {rotatorTypeColumnName}",msSqlConnection)){
            using (SqlDataReader reader = getMaxSequenceByRotatorTypeQuery.ExecuteReader()){
                while (reader.Read()){
                    result.Add(new RotatorEntry{
                        RotatorType = reader[rotatorTypeColumnName].ToString(),
                        MaxSequence = Convert.ToInt32(reader["MaxSequence"])
                    });
                }
            }
        }

        return result;
    }


}

//Figure out what to do when you cant add 1 anymore to the sequence like for exaple there is only 8 raids what happens when you add 1 to 8. It will most likely fail. See what the error is and catch it. 
//Add more logging to see what code is doing. 

File: WeeklyRotators/update_active_weekly_rotators.py
import mssql_python
from dotenv import load_dotenv
import os
import pandas as pd
import json

# Load Config file
config = load_dotenv("REDACTED")

# Extract DB connection details
db_host = os.getenv("DB_HOST")
db_username = os.getenv("DB_USERNAME")
db_password = os.getenv("DB_PASSWORD")
db_database = os.getenv("DB_PROD")

# Table we will be updating
active_weekly_rotators_table_name = 'ActiveWeeklyRotatorsTable'

def connect_to_db():
    connection_string = (
        f"Server={db_host};"
        f"Database={db_database};"
        f"Uid={db_username};"
        f"Pwd={db_password};"
        "Encrypt=Yes;"
        "TrustServerCertificate=Yes;"
    )
    print(f"Connecting to {connection_string}")
    connection = mssql_python.connect(connection_string)
    cursor = connection.cursor()
    return connection, cursor

def check_table_exists(cursor, table_name):
    cursor.execute(f"SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = '{table_name}'")
    does_table_exist = cursor.fetchone()
    return bool(does_table_exist)

def check_row_exists(cursor, table_name):
    cursor.execute(
        f"SELECT 1 FROM dbo.{table_name} WHERE Name = 'ActiveWeeklyRotators'"
    )
    return bool(cursor.fetchone())


def create_and_populate_active_weekly_rotators_table(connection, cursor, table_name):
    # Drop ActiveWeeklyRotatorsTable if it exists
    print('Dropping ActiveWeeklyRotatorsTable to recreate it')
    cursor.execute(f"DROP TABLE IF EXISTS dbo.{table_name};")
    connection.commit()

    # Get all the activities with sequence 1
    cursor.execute(f"SELECT Hash FROM RotatorSchedule WHERE Sequence = 1")
    rotator_schedule_table = cursor.fetchall()

    # Convert table to df
    rotator_schedule_df = pd.DataFrame( rotator_schedule_table, columns=["Hash"])

    # Extract Hash list from df and convert it to a string
    default_rotator_schedule_hash_list = rotator_schedule_df["Hash"].tolist()
    default_rotator_schedule_hash_string = json.dumps(default_rotator_schedule_hash_list)
    print(default_rotator_schedule_hash_string)

    # Execute the Create Table Query and Populate it
    print('Creating ActiveWeeklyRotatorsTable')
    cursor.execute(f"CREATE TABLE dbo.{table_name} (Name VARCHAR(MAX) , RotatorList VARCHAR(MAX)); INSERT INTO dbo.{table_name} (Name, RotatorList) VALUES ('ActiveWeeklyRotators', '{default_rotator_schedule_hash_string}');")
    connection.commit()

def update_active_weekly_rotators():
    # Connect to DB
    connection, cursor = connect_to_db()

    # Check if Active Weekly Rotators Table Exist and Create a Populate it if it does not
    active_weekly_rotators_table_status = check_table_exists(cursor, active_weekly_rotators_table_name)
    if not active_weekly_rotators_table_status:
        print("No active weekly rotators table found.")
        create_and_populate_active_weekly_rotators_table(connection=connection, cursor=cursor, table_name=active_weekly_rotators_table_name)
    else:
        print("Active weekly rotators table found.")

    # Check if proper columns exist within Active Weekly Rotators Table
    check_active_weekly_rotators_row_status = check_row_exists(cursor=cursor, table_name=active_weekly_rotators_table_name)
    if not check_active_weekly_rotators_row_status:
        print("No active weekly rotators table found.")
        create_and_populate_active_weekly_rotators_table(connection=connection,cursor=cursor, table_name=active_weekly_rotators_table_name)
    else:
        print("ActiveWeeklyRotators row found")

    new_rotator_schedule_hash_list = []
    if active_weekly_rotators_table_status and check_active_weekly_rotators_row_status:
        # Grab list from ActiveWeeklyRotatorsTable
        cursor.execute(f"SELECT * FROM dbo.{active_weekly_rotators_table_name}")
        active_weekly_rotators_table = cursor.fetchall()
        active_weekly_rotators_df = pd.DataFrame(active_weekly_rotators_table, columns=["Name", "RotatorList"])
        rotators_list_string = active_weekly_rotators_df["RotatorList"].iloc[0]
        active_rotators = json.loads(rotators_list_string)

        # Get the Rotator Schedule
        cursor.execute(f"SELECT * FROM RotatorSchedule")
        rotator_schedule_table = cursor.fetchall()
        rotator_schedule_df = pd.DataFrame(rotator_schedule_table, columns=["Hash", "RotatorType", "Sequence", "Name"])
        print(rotator_schedule_df)


        for active_rotator in active_rotators:

            print(active_rotator)
            row = rotator_schedule_df[rotator_schedule_df["Hash"] == active_rotator].iloc[0]
            hash = row["Hash"]
            rotator_type = row["RotatorType"]
            sequence = row["Sequence"]
            name = row["Name"]
            print(name)

            # Get the max Sequence number for the current rotator
            cursor.execute(f"SELECT MAX(Sequence) from RotatorSchedule WHERE RotatorType = '{rotator_type}'")
            max_rotator_sequence = cursor.fetchone()[0]

            if (sequence + 1) > max_rotator_sequence:
                next_sequence = 1
            else:
                next_sequence = sequence + 1

            next_rotator_row = rotator_schedule_df[(rotator_schedule_df["Sequence"] == next_sequence) & (rotator_schedule_df["RotatorType"] == rotator_type)].iloc[0]
            new_rotator_schedule_hash_list.append(int(next_rotator_row["Hash"]))
        print(active_rotators)
        print(new_rotator_schedule_hash_list)
        print(type(new_rotator_schedule_hash_list))

    raid = False
    dungeon = False
    exotic_mission = False

    # Check if one of each type of rotator exist. This is important for the website as it expects a list of index 3 and one for each type
    for new_rotator_schedule_hash in new_rotator_schedule_hash_list:
        row = rotator_schedule_df[rotator_schedule_df["Hash"] == new_rotator_schedule_hash].iloc[0]
        rotator_type = row["RotatorType"]

        if rotator_type.lower() == "dungeon":
            dungeon = True
        elif rotator_type.lower() == "exotic mission":
            exotic_mission = True
        elif rotator_type.lower() == "raid":
            raid = True

    if raid == True and dungeon == True and exotic_mission == True and len(new_rotator_schedule_hash_list) == 3:
        new_rotator_schedule_hash_string = json.dumps(new_rotator_schedule_hash_list)
        print(new_rotator_schedule_hash_string)
        print(type(new_rotator_schedule_hash_string))
        cursor.execute(
            f"UPDATE dbo.{active_weekly_rotators_table_name} SET RotatorList = ? WHERE Name = ?",
            (new_rotator_schedule_hash_string, "ActiveWeeklyRotators")
        )
        connection.commit()







update_active_weekly_rotators()







